common/ledger/blkstorage/cache.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package blkstorage
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+)
+
+const (
+	estimatedBlockSize = 512 * 1024
+)
+
+type Cache struct {
+	cacheLock    sync.RWMutex
+	disabled     bool
+	cache        map[uint64]cachedBlock
+	sizeBytes    int
+	maxSeq       uint64
+	maxSizeBytes int
+}
+
+type cachedBlock struct {
+	block     *common.Block
+	blockSize int
+}
+
+func NewCache(maxSizeBytes int) *Cache {
+	isCacheDisabled := maxSizeBytes == 0
+
+	return &Cache{
+		disabled:     isCacheDisabled,
+		cache:        make(map[uint64]cachedBlock, maxSizeBytes/estimatedBlockSize),
+		maxSizeBytes: maxSizeBytes,
+	}
+}
+
+func (c *Cache) retrieveCachedBlock(seq uint64) (*common.Block, bool) {
+	if c.disabled {
+		return nil, false
+	}
+
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+
+	cachedBlock, exists := c.cache[seq]
+	return cachedBlock.block, exists
+}
+
+func (c *Cache) cacheBlock(block *common.Block, blockSize int) {
+	if c.disabled {
+		return
+	}
+
+	seq := block.Header.Number
+
+	if c.maxSeq > seq {
+		return
+	}
+
+	// Insert the block to the cache
+	c.maxSeq = seq
+
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+
+	c.sizeBytes += blockSize
+
+	c.cache[seq] = cachedBlock{block: block, blockSize: blockSize}
+
+	// If our cache is not too big, return early and don't evict an entry
+	if c.sizeBytes <= c.maxSizeBytes {
+		return
+	}
+
+	c.evictLowestCachedBlock()
+}
+
+func (c *Cache) evictLowestCachedBlock() {
+	cachedItemCount := len(c.cache)
+
+	// Given a series of k > 0 consecutive elements: {i, i+1, i+2, ... , i+k-1}
+	// If the max sequence is j then j=i+k-1, and then the lowest element i is j-k+1
+	evictedIndex := c.maxSeq - uint64(cachedItemCount) + 1
+	evictedBlock, exists := c.cache[evictedIndex]
+	if !exists {
+		panic(fmt.Sprintf("programming error: last stored block sequence is %d and cached block count"+
+			" is %d but index to be evicted %d was not found", c.maxSeq, cachedItemCount, evictedIndex))
+	}
+	delete(c.cache, evictedIndex) // Delete minimum entry
+	c.sizeBytes -= evictedBlock.blockSize
```

Comment:
When removing a single block from the cache, it is assumed that their sizes are comparable. 
If they are not? 
If one block is much larger than the others?
Perhaps more than one block needs to be removed?

Reply:
I decided so because you can only remove one block from the cache per block added to the cache.
And the number of blocks in the cache can only increase or stay the same.
If there are 10 blocks in the cache now, it will never become 5 if the chain is not interrupted.

Reply:
illustrate:
let's say max 100.
add 100 blocks with weight 1.
occupied 100

add a block with a weight of 95
total 100+95-1 = 194
We'll say it 100 times.
and with a maximum of 100, the sum of the weights would be 9,500.

it's all in your algorithm.

Reply:
I see your point. I don't think it's an issue unless the system increases its block size over time. 

Regardless, if we increase the block size we get to a situation where the cache's max size is not taken into consideration.

I changed accordingly, thanks for pointing this out.